### DISCUSSION TITLE
Support for the new Sonoff POWCT

### DISCUSSION THREAD

**Discussion by veno77** [arendst/Tasmota#21131] [Q&A] [answered]
Hello,
Sonoff have new POW Ring Smart Power Meter - POWCT based on ESP32. When can we expect template in the firmware?
Thanks,
Veno

**Comment by barbudor** [comment]
When you will have purchased one and teared it open to find out the connexion

The team doesn't have the budget to purchase each and every new device in the market, community has to do its part of the job

**Comment by rothlive** [comment]
I have purchased one.
I could test a new firmware.

![IMG20240626151814](https://github.com/arendst/Tasmota/assets/10054386/b877d4eb-e3b1-4757-a51c-64e7c16b26ce)
![IMG20240626152101](https://github.com/arendst/Tasmota/assets/10054386/3e2ed937-f78a-49a6-9e43-a7bd716043aa)
![IMG20240626152241](https://github.com/arendst/Tasmota/assets/10054386/9b7e1b35-457b-4beb-9941-75907833106e)

**Reply by vivek4444** [reply]
i also got Sonoff pow ring but not able to perform auto config with latest tasmota release .
Can you please share GPIO mapping .

**Comment by arendst** [comment]
From the pictures it seems to me all hardware is already supported by tasmota.  All you have to do is finding out which esp32 gpio connects to what chip. Then make the template.

**Reply by arendst** [reply]
For testing perform the following:

- Load default ``tasmota32.bin`` binary supporting all needed hardware.
- Load the Sonoff THR16 template and check if the display works. If not try next
- Load the Sonoff POWR316 template and check if the display works.
- Find the correct GPIOs connected to ``CSE7761 Rx`` and ``CSE7761 Tx``

**Reply by rothlive** [reply]
How can I find out?
With https://templates.blakadder.com/sonoff_THR316D.html
Display -- comes on briefly at startup. Then goes off again
Relay switch works.
GPIO21 | Relay 1




</body>
</html>

https://templates.blakadder.com/sonoff_POWR320D.html
Nothing works.

**Reply by rothlive** [reply]
CSE7761 Rx CSE7761 Tx find.
Only enter everything once from GPIO00 to 39 or is there a system for finding this?

**Reply by arendst** [reply]
"simply" follow the print copper wires from the cse7761serial connection to the esp32 gpio. As the display is connected like a thr316 you might want to chk gpio25 and 26 first being the "normal" thr316d sensor gpios

**Reply by rothlive** [reply]
"simply" follow the print copper wires from the cse7761serial connection to the esp32 gpio.

I don't know that.
I'm not very familiar with the circuit board.
Can't you see that from the pictures?

**Comment by tiagofreire-pt** [comment]
From the _neighbor next door_, all the GPIO configuration: https://devices.esphome.io/devices/sonoff-powct/

**Reply by rothlive** [reply]
Thank you !

**Comment by arendst** [comment]
So using the latest dev branch (v14.1.0.3) and this template:

```
{"NAME":"Sonoff POWCT","GPIO":[32,0,0,0,0,9280,0,0,0,320,0,576,0,9184,9216,0,0,224,0,9248,0,7329,7296,0,0,0,0,0,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":1}
```

It should work.

NB. Latest dev branch supports single channel CSE7761 when gpio is ``CSE7761 Rx2`` as configured in above template.

**Reply by rothlive** [reply]
Unfortunately I cannot find the dev branch (v14.1.0.3) for the ESP32.
Only for the ESP8266 at http://ota.tasmota.com/tasmota/.
But there is an ESP32 installed for it I can only find the version 14.1.0 Rachel
http://ota.tasmota.com/tasmota32/release/

**Reply by rothlive** [reply]
![Unbenannt000](https://github.com/arendst/Tasmota/assets/10054386/05540bce-b381-4427-8cdb-67b075cadbd4)
Display not works at start only --- and then nothing.
Switch on off works.
With : `{"NAME":"Sonoff POWCT","GPIO":[32,0,0,0,0,9280,0,0,0,320,0,576,0,9184,9216,0,0,224,0,9248,0,7329,7296,0,0,0,0,0,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":1}`

**Reply by arendst** [reply]
The template only works with latest dev. Follow the links at the top of the ota.tasmota.com page to find latest dev.

**Reply by rothlive** [reply]
![Unbenannt000](https://github.com/arendst/Tasmota/assets/10054386/f7edcf9c-a28c-4202-b6e4-fab3a9307a62)
ok I have now found it.
The display is still not working.
Only when starting --- and then it goes off.

**Reply by arendst** [reply]
I see. I need to make some changes to the display driver.

For the short term you might want to use ``DspLine`` commands. I'm not sure what the sensor output is for this PowCt but if it look slike this:
```
    //      {"ENERGY":{"TotalStartTime":"2022-07-05T16:01:39","Total":0.000,"Yesterday":0.000,"Today":0.000,"Power":0.00,"ApparentPower":0.00,"ReactivePower":0.00,"Factor":0.00,"Frequency":50,"Voltage":231.7,"Current":0.000}}
    // index: 1         2                                      3             4                 5             6            7                    8                    9             10             11              12
    //  unit: 0         0                                      4 (kWh)       4 (kWh)           4 (kWh)       4 (W)        0 (VA)               0 (VAr)              0             0              3 (V)           3 (A)
```

you could try commands ``dspline1 11,3,3,4`` and ``dspline2 12,3,6,4``

**Reply by rothlive** [reply]
![Unbenannt000](https://github.com/arendst/Tasmota/assets/10054386/1f6c6026-d5d5-4e92-b744-dc94dc96474b)
dspline1 11,3,3,4 and dspline2 12,3,6,4 works fine.
Display works perfect.

**Reply by rothlive** [reply]
Then you can make a new Tasmota Relase with support for the Sonoff POWCT.

**Reply by arendst** [reply]
Try latest dev from 16:51 which contains the display fix. No need for ``DspLine`` anymore.

Next release will be in mid August.

**Reply by rothlive** [reply]
Ok
This is not necessary even after restarting the display remains functional. I don't have to enter the command again

**Reply by rothlive** [reply]
I am very pleased that the Sonoff now works with Tasmota.
Many thanks for your help.

**Comment by ajaypala** [comment]
Does Tasmota support the 2nd current input channel on the CSE7761? I'm wondering if I can add C37, R16, R11, C36, J8 and connect a 2nd Current ring sensor.

**Reply by arendst** [reply]
The second current ring only makes sense as the ct is used on the same phase. The cse7761 driver already supports two channels like the dualr3. The display driver supports the second channel with dspline commands. Good luck.

**Reply by tiagofreire-pt** [reply]
How could be used that second display driver channel? The dspline command is highly undocumented. Could you help?

**Reply by arendst** [reply]
```
 * DspSpeed 2..127 = Display rotation speed in seconds if more than one value is requested
 * DspLine<1|2> <index>,<unit>,<index>,<unit>,... = Display specific JSON value and rotate between them
 *   unit 0 = None
 *        1 = Temperature (Line 1 only)
 *        2 = %RH (Line 2 only)
 *        3 = Both V (Line 1 only) / A (Line 2 only)
 *        4 = Both kWh (Line 1 only) / W (Line 2 only)
 * DspLine1 0 and DspLine2 0 = Default of temperature/humidity
 * DspLine9 = Show sensor data
 *
 * Example: {"SCD30":{"CarbonDioxide":746,"eCO2":727,"Temperature":30.6,"Humidity":43.6,"DewPoint":16.8}}
 *   index:   1        2                   3          4                  5               6
 *    unit:   0        0 (ppm)             0 (ppm)    1 (C or F)         2 (%RH)         1 (C or F)
 *
 *   DspLine1 4,1,3,0 = Temperature and eCO2
 *   DspLine2 2,0,5,2 = CarbonDioxide and humidity
```

**Comment by arendst** [comment, marked as answer]
Latest dev v14.1.0.3 provides an updated version of the CSE7761 energy driver used by Sonoff POWCT. The template has also been finalized as being 
```
{"NAME":"Sonoff POWCT","GPIO":[32,0,0,0,0,9280,0,0,0,288,0,576,0,9184,9216,0,0,224,0,9248,0,7329,7296,0,0,0,0,0,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":1}
```

The configuration is also available as Auto-configuration.

Using a kind of "reverse engineering" the correct CSE7761 init for Sonoff POWCT has been recorded using below set-up:
![29F067E0-0AEA-4071-AE6B-34F67509531A](https://github.com/arendst/Tasmota/assets/11044339/a7fc255c-f71f-4cb4-880f-2f36d66a895e)

**Reply by antoniocanelas** [reply]
Hi @arendst, thank you for sharing this info. Already updated many different sonoff components with tasmota firmware, latest were the POWR316D and POWR320D. Now I'm trying to update the one you have on the photo "Sonoff POW Ring | POWCT". Looking to this photo I can't understand exactly what are the pins that you use and were are connected to FT232RL. Is it possible to someone of you give me details? Thank you in advance.

**Reply by pluk77** [reply]
@antoniocanelas  The pins are not on the board yet. I flashed one recently and had to solder header pins to the holes in the PCB. You can see the holes in the board on the photo next to the chip. The above photo is of the bottom side of the PCB, so you will have to unscrew the board and remove it from the housing to properly solder on the pins. Once done, you can put it back in the case and the pins are then accessible to connect to the FT232RL. Hope this helps

**Reply by sfromis** [reply]
If what you ask about are the physical pins (SP10 etc) used to connect to the clearly labled spots on the PCB, that is a product from SensePeek named "PCBite". In this case, I'd actually prefer to solder a small header to the 4 easy-access holes, and leave it there. Still, PCBite is a nice product, albeit on the pricey side. 

Also, they are probably **not** connected to a FT232L, but another USB/serial adapter. Most of those "FT232L" devices are fakes unable to supply the power for reliably flashing ESP32 devices.

**Reply by sfromis** [reply]
BTW. When I say "clearly labeled", I'm referring to this crop from the image above: 
![image](https://github.com/user-attachments/assets/102a0dbd-11b5-49e4-a21b-f4107017001f)

**Reply by antoniocanelas** [reply]
Thank you for your support @pluk77 and @sfromis for your support. I'll try this weekend.

Btw on the others sonoff devicesI flashed the firmware with this, it works to this sonoff pow ring as well right?
![WhatsApp Image 2024-11-26 at 20 08 36](https://github.com/user-attachments/assets/ce309017-a7d1-43a2-9199-f41e5ddfc9da)

**Reply by sfromis** [reply]
That specific USB/serial adapter is a main reason why I wrote:
> Most of those "FT232L" devices are fakes unable to supply the power for reliably flashing ESP32 devices.

Better success rate with ESP8266, but ESP32 needs more power; thus a bit of a gamble. Of course, also remember gpio 0 grounded before/while applying power.

**Reply by antoniocanelas** [reply]
Hi thank you  again for the reply, just to be sure. I'll buy the ESP8266, is this one?
![image](https://github.com/user-attachments/assets/257b4be2-a44b-478b-991e-5d57a07f56da)

Do you have any reference, just to be sure that I will buy the right one?

Thank  you for your help

**Reply by sfromis** [reply]
It's not so much about ESP8266, as using a NodeMCU board as USB/serial adapter requires disabling the MCU, to connect the serial adapter chip to external serial instead, where you need a driver to support that chip, CH340G in the above image. And those NodeMCU boards do generally have a 500 mA AMS1117 (in your image below the yellow ones), with good power for flashing. Yes, the one in the image should be ok to use.

**Reply by antoniocanelas** [reply]
Thank you so much, already able to flash Sonoff POW Elite.

**Reply by antoniocanelas** [reply]
Regarding POW Sonoff  Ring POWCT, I'm allways getting this error:
![image](https://github.com/user-attachments/assets/a9c846fb-754d-4c92-990e-a29ec3fdc543)

**Reply by sfromis** [reply]
Typical reasons include wrong wiring, not grounding gpio 0 before/while applying power, or insufficient power from the USB/serial adapter.

**Reply by antoniocanelas** [reply]
![image](https://github.com/user-attachments/assets/8dac69f9-6507-4cfe-93ce-b487b817f964)
Error also on  esptool.py.

**Reply by antoniocanelas** [reply]
Is it normal to have display on during the proccess. Already tried more that  20 times, make exactly  the same on pow elite. Use more  than  one powct. Changed  the cables, and continues  working on pow  elite but not on   powct. Thank  you

<img width="672" alt="image" src="https://github.com/user-attachments/assets/97985c5f-099d-427d-aab7-7a018c6a9301">

**Reply by antoniocanelas** [reply]
Ok  solved :) Able to flash connecting the cable immediately after click  on the flash,  with   the button pressed during that time.

Thank you

**Reply by vivek4444** [reply]
Auto config is not working in tasmota can i please get Module configuration which show mapping of GPIO
@arendst

**Reply by sfromis** [reply]
You can see the actions from autoconf here, which is mainly the template with gpio configs. Better apply that, instead of Configuration/Module.
https://github.com/tasmota/autoconf/blob/main/raw/esp32/Sonoff_POWCT/init.bat

**Comment by tiagofreire-pt** [comment]
Hi, @arendst 

On dev 14.1.0.3 and dev 14.1.0.4, the instantaneous power at the display seems to be with scientific notation.

![Screenshot_20240804_130532_Firefox Focus](https://github.com/user-attachments/assets/00000376-99e2-4370-b957-4e148ff615d6)
![20240804_130655](https://github.com/user-attachments/assets/44fc835a-3cfd-4bdd-8849-5ea2381d0437)
![20240804_123222](https://github.com/user-attachments/assets/47aa7726-e93f-478b-a193-2e186d9909ed)

**Reply by tiagofreire-pt** [reply]
Is this a bug?

**Reply by arendst** [reply]
Not according to the changelog:
```
- ESP32 TM1621 number overflow from "9999" to "12E3" (#21131)
```

Although it shouldn't do it for values below 9999.... Can you find out what the non scientific number would have been? Is it 4000 or 40000?

EDIT: I see. It is 3619.1 rounded to 40E2. Not good. I'll see how to get rid of decimals first before showing scientific numbers

**Reply by tiagofreire-pt** [reply]
Thanks for your kind support.

I recommend that the device should show on the display, using scientific notation, for power values above 9999 W, as it's limited to 4 digits. And the device supports up to 23000 Watt (23 kW - 23E3 W)

For lower than 1000 W, would be awesome to get with a decimal, such as 66.7 or 999.9 W. Is this possible?

**Reply by tiagofreire-pt** [reply]
For both upper examples shown of the issue:

36xx W -》36E2 W
40xx W -》40E2 W

**Reply by arendst** [reply]
Try latest dev branch where it is fixed now.

Thx for reporting.

**Reply by tiagofreire-pt** [reply]
Works flawlessly now! Thanks!

**Comment by tiagofreire-pt** [comment]
Could help on why this meter does not count and measure low power loads, about 5W, displaying 0W?

Is there a setting I'm missing on Tasmota?

**Reply by arendst** [reply]
As the CT provided is for 100A my guess it won't detect a load below 5W. You'll need a smaller CT for that (which sonoff doesn't provide).

**Reply by tiagofreire-pt** [reply]
Could you confirm if it's a driver limitation? Because the manufacturer states a range of 0-100A.

**Reply by tiagofreire-pt** [reply]
Not even setoption39 to 0 makes a difference.

**Comment by arendst** [comment]
As said, it's a hardware limitation. Also as said by the manufacturer, it supports from 0 to 100A. They don't tell you it cannot detect current changes in the milliamperes needed to detect changes below 5W Hence the existence of 10A, 20A, and 50A CT's.

Search google and learn the CT functionality.

**Reply by tiagofreire-pt** [reply]
Is it possible to adjust the sensibility of the ADC to others CT, with Tasmota?

**Reply by arendst** [reply]
It probably is possible. See below excerp from the CSE7761 driver:
```
/*
    System Control Register (SYSCON)  Addr:0x00  Default value: 0x0A04
    Bit    name               Function description
    15-11  NC                 -, the default is 1
    10     ADC2ON
                              =1, means ADC current channel B is on (Sonoff Dual R3)
                              =0, means ADC current channel B is closed (Pow CT)
    9      NC                 -, the default is 1.
    8-6    PGAIB[2:0]         Current channel B analog gain selection highest bit
                              =1XX, PGA of current channel B=16 (Sonoff Dual R3)
                              =011, PGA of current channel B=8
                              =010, PGA of current channel B=4
                              =001, PGA of current channel B=2
                              =000, PGA of current channel B=1 (Pow CT)
    5-3    PGAU[2:0]          Highest bit of voltage channel analog gain selection
                              =1XX, PGA of voltage U=16
                              =011, PGA of voltage U=8
                              =010, PGA of voltage U=4
                              =001, PGA of voltage U=2
                              =000, PGA of voltage U=1 (Sonoff Dual R3 / Pow CT)
    2-0    PGAIA[2:0]         Current channel A analog gain selection highest bit
                              =1XX, PGA of current channel A=16 (Sonoff Dual R3)
                              =011, PGA of current channel A=8
                              =010, PGA of current channel A=4
                              =001, PGA of current channel A=2
                              =000, PGA of current channel A=1 (Pow CT)
*/
    if (CSE7761_MODEL_POWCT == CSE7761Data.model) {
//      Cse7761Write(CSE7761_REG_SYSCON | 0x80, 0x0A00);  // Pow CT
      Cse7761Write(CSE7761_REG_SYSCON | 0x80, 0xFE00);  // Pow CT - Tasmota (enable B)
    } else {      
      Cse7761Write(CSE7761_REG_SYSCON | 0x80, 0xFF04);  // Sonoff Dual R3
    }
```
Notice the current gain difference between the dualR3 and the CT.

If you compile your own, play with the value and see if it solves your low power issue (and probably introducing unavailable 100A detection).

**Reply by tiagofreire-pt** [reply]
If those gains could be set dynamically, through a setoption on tasmota, such as the setoption39, would be awesome. Because compiling the code is not a friendly user way and not practical if you want to remain updated with the latest tasmota release versions.

Do you mind taking a closer look into that?

**Reply by arendst** [reply]
No unless it makes sense. The current settings are the provided ones from Sonoff to be used with their hardware. Adding user support for changing registers opens a can of worms as there are many options to consider and most of them have no impact at all.

I suggest you make some changes to the current gain and see if it makes sense at all. Again, there won't be many users wanting this except the video kids which seem te be unable to think and just try changes without understanding what they're doing.

**Reply by tiagofreire-pt** [reply]
If possible, at least, within the dev version as a debug tool to find a better value.

**Comment by pluk77** [comment]
Thank you so much for the hard work. 

For those wanting to do the same:

I flashed my POWCT using the web installer using Edge. It worked but it could not complete the installation. In the browser console I saw the following message coming up: 

   `Improv initialization failed. Error: Improv Wi-Fi Serial not detected`

I then cycled the power and had to connect to the wifi of the device to setup the wifi. (All by providing power via Serial Programmer / USB, not mains)

Thereafter I used the auto-configuration option, selected POWCT and the device is working!

Device info according to esptool:

```
   Chip is ESP32-D0WD-V3 (revision 3)
   Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
   Crystal is 40MHz
   MAC: e8:6b:ea:f4:ad:34
```

**Reply by Kiogora** [reply]
I note the frequency in the CSE7761 driver is an integer. Where can I modify in order to have these as float values reported over MQTT? Should I start at the driver or is there a way to achieve this on the UI.

**Reply by arendst** [reply]
Start at reading the docs. Try command `FreqRes 2`

**Comment by bT0mm** [comment]
Are the unused GPIOs available for external sensors (I2C, SPI...) ?
With other Sonoff power monitoring devices without CT they can't be used because they're connected to mains...

**Comment by jaggedjagged** [comment]
Hi,
can anybody help me?
I flashed the POWCT with Tasmota ESP32-DevKit.
Connection via WLAN worked fine.
When i choose CSE7761 for IO GPIO0 nothing shows up.
Display is dead as well.
So how to configure that? Or did i choose the wrong binary?
Thanks a lot!

**Reply by sfromis** [reply]
Instead of configuring each individual gpio, the recommendation in this thread is to perform auto-configuration (selecting the device in Configuration / Auto-conf), to apply a full template in one go.

And that is nothing like "CSE7761 for IO GPIO0" which is not going to work.

**Comment by sarahschreiber** [comment]
Hi,

I'm trying to connect a DS18B20 temperature sensor to a Sonoff POW CT using the serial interface pins (GPIO1 or GPIO3). However, the temperature readings are completely inaccurate.

Is there a known reason why this might happen? Could it be related to the hardware design of the device, or is the DS18B20 simply not supported on these pins?

Any advice or experience with this setup would be greatly appreciated.

Thanks,

Sarah

**Reply by arendst** [reply]
As the POW-CT connects it's serial interface to the line voltage we do not support any external connection other than the supported port ones

**Reply by sarahschreiber** [reply]
Is this different from the POWR316D? It seems to work fine on the POW R3. Or is it just a coincidence that the temperature readings appear to be correct?
